src/lsst/cmservice/common/graph.py

```diff
@@ -192,3 +199,252 @@ def processable_graph_nodes(g: nx.DiGraph) -> Iterable[Node]:
+    namespace: UUID,
+    session: AsyncSession | None = None,
+    commit: bool = True,
+) -> None:
+    """Apply an append operation to a graph by adding a new node_1 parallel
+    to an existing node_0 where all adjacencies are copied to node_1.
+
+    ```
+    A --> B --> C  becomes  A --> B --> C
+                              `-> X -/
+    ```
+
+    This behavior is appropriate for the common use case of adding group nodes
+    in parallel to an "anchor" group node, such as the case where "B" was added
+    to the graph as an "insert" and it is known that additional nodes must be
+    parallel to it.
+
+    Different contexts with different kinds of "anchor" nodes for this op
+    require a different algorithm, for example the case of adding a "Step" node
+    in parallel with an existing "Step" node that has already prepared its
+    groups -- the new node should not form adjacencies with those groups!
+
+    ```
+    A --> B --> B1 --> Bx --> C
+            `-> B2 -/
+    ```
+
+    In this case, appending a node X to B where Bn are its groups and Bx is its
+    collect-step, X should be adjacent to C, not to B1 and B2:
+
+    ```
+    A ---> B --> B1 ---> Bx --> C
+      `      `-> B2 -/      /
+       `-> X --------------/
+    ```
+
+    In this case, when determining the "downstream edges" to inherit for the
+    new node X, nodes of kind "group" and "collect" should be invisible to the
+    algorithm.
+
+    Other contexts may be unsupported or meaningless to the application, in
+    which case the append request should abend with an error or a warning along
+    with a no-op.
+
+    For example, appending a node to a "collect-step" is not a meaningful oper-
+    ation because there should only ever be one node adjacent to a set of group
+    nodes. Likewise, no node should ever be made parallel to a "start" or "end"
+    node as this would create an invalid graph. Future node kinds, for instance
+    a "breakpoint" or "approval" node, would likewise not support append oper-
+    ations as this would give the graph a way to bypass an intentional check-
+    point.
+
+    Raises
+    ------
+    NotImplementedError
+        Raised if the node parallelization context is not supported or not yet
+        implemented. It is left to the caller to decide how to proceed.
+    """
+    if TYPE_CHECKING:
+        assert session is not None
+
+    # define a context for this operation based on the node_0 kind.
+    n_0 = await session.get_one(Node, ident=node_0)
+
+    match n_0.kind:
+        case ManifestKind.group | ManifestKind.other:
+            # create new "downstream" edges with node_1
+            s = select(Edge).with_for_update().where(Edge.source == node_0)
+            downstream_edges = (await session.exec(s)).all()
+        case ManifestKind.grouped_step:
+            # node_1 should become adjacent to the next node in node_0's path
+            # that is not a group or a collect node.
+            raise NotImplementedError
```

Comment:
If I'm understanding this right, this is avoiding the complicated cases of finding the downstream anchor by declaring those not supported? Worth saying that in the docstring.

Reply:
For this PR, this case is present as a placeholder, the implementation is in DM-52232.